engine/merge_accessors.ml

```diff
@@ -0,0 +1,54 @@
+open Ast
+
+type constraint_tree =
+  | Failure
+  | Leaf of Ast.target_blackbox
+  | Node of (pi * constraint_tree) list * (pi list * constraint_tree) option
+and
+  pi = { var: sym_value; op: Target_sym_engine.piop }
+and
+  sym_function = variable * constraint_tree
+and
+  sym_catch = exitpoint * variable list * constraint_tree
+and
+  sym_value =
+  | AcRoot of variable
+  | AcField of sym_value * int
+  | AcTag of sym_value * int
+
+let rec merge : Target_sym_engine.constraint_tree -> constraint_tree =
+  let rec subst_pi (pi: Target_sym_engine.pi) : pi =
+    match pi.var with
+    | AcAdd (svalue, a) -> let op':Target_sym_engine.piop =  match pi.op with
+        | Tag i -> Tag (a+i)
+        | NotTag i -> NotTag (a+i)
+        | Int i -> Int (a+i)
+        | NotInt i -> NotInt (a+i)
+        | Ge i -> Ge (a+i)
+        | Gt i -> Gt (a+i)
+        | Le i -> Le (a+i)
+        | Lt i -> Lt (a+i)
+        | Eq i -> Eq (a+i)
+        | Nq i -> Nq (a+i)
+        | Isout i -> Isout (a+i)
+        | Isin i -> Isin (a+i)
+      in
+      subst_pi {var=svalue; op=op'}
+    | AcRoot v -> {var=AcRoot v; op=pi.op}
+    | AcField (s, i) ->
+      let inner = subst_pi {var=s; op=pi.op} in
+      {var=AcField (inner.var, i); op=inner.op}
+    | AcTag (s, i) ->
+      let inner = subst_pi {var=s; op=pi.op} in
+      {var=AcTag (inner.var, i); op=inner.op}
```

Comment:
The definitions in the `AcField` and `AcTag` case do not really make sense, for example they will let you translate the  constraint `(root+3).0 = 2` (technically, `{var = AcField(AcAdd(AccRoot, 3), 0) pi = Eq 2}`) into the constraint `root.0 = 5`, and the two are not at all equivalent (technically we *could* give a meaning to `(root+3).0` as dereferencing the result of a pointer arithmetic, and then you see that the rewrite is bogus).

I would prefer to just assume that there cannot be an `AcAdd` case under an `AcField` or an `AcTag`, and not rewrite recursively under them. (It is good that there is a recursive case in the `AcAdd` case, as you may have several additions on top of each other.)

Reply:
I agree. Originally I wrote the code this way because in any case the AcField, AcRoot and AcTag need to be transformed from Target_sym_engine.pi to Merge_accessors.pi anyway. Now it is more explicit